Prep for Docker log drivers announcement

tgross opened the pull request:
@misterbisson I'm opening this PR but it's still work-in-progress.
The current master branch prior to merging this was tagged as `preview`.

Changes:
- check package signatures for Logstash and Containerbuddy during build
- discover Consul via Triton CNS when deployed to Triton (but works locally via envvar)
- consolidate the Docker Compose files to take advantage of getting rid of `--link`s in the main file
- split running the stack (which can be done with just `docker-compose up -d`) from testing the stack
- pull "assert the world" code from Triton Touchbase blueprint into the `./test.sh check` function
- lots of README updates

tgross committed:
Consolidate compose files, discover Consul via Triton CNS

- Drop common-compose.yml by moving its services into docker-compose.yml.
- Update local-compose.yml to extend the main (on-Triton) docker-compose.yml
  file instead.
- Use an environment variable to pass the Triton CNS svc location for Consul
  to Containerbuddy.
- Upgrade Containerbuddy and check SHA1 hash after download

tgross committed:
Switch to Alpine Linux for Kibana and Logstash

tgross committed:
MOAR AUTOPILOT: move index bootstrapping into Kibana container

tgross committed:
Split out running the stack vs testing the stack

- rename start.sh to test.sh
- move the Nginx services to their own compose file
- folded common.sh into the test rig because we only have one place of use

tgross committed:
Make `check` in test script more robust

tgross committed:
Use Triton CLI in test.sh

tgross committed:
Move Kibana and Logstash back to Debian from Alpine

tgross committed:
Force ES to listen on private IPs only

tgross committed:
Assign ES master+data node to advertise itself as a master

tgross committed:
Update README to match current implementation

tgross commented:
@misterbisson this and https://github.com/tgross/triton-elasticsearch/pull/2 are probably ready for re-review at this point.

tgross committed:
Data center, not data-center

misterbisson commented on README.md:
```diff
@@ -1,76 +1,106 @@
+- [Kibana](https://www.elastic.co/products/kibana), as a web UI to search the logs we store in Elasticsearch.
+- [Consul](https://www.consul.io/), acting a service catalog to support discovery
+- [Containerbuddy](http://containerbuddy.io), to help with service discovery and bootstrap orchestration
+- [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform, including the new [Docker log drivers](https://www.joyent.com/blog/docker-log-drivers) support.
```
Full stop with just the following, maybe?

> [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform

misterbisson commented on README.md:
```diff
@@ -1,76 +1,106 @@
+- [Consul](https://www.consul.io/), acting a service catalog to support discovery
+- [Containerbuddy](http://containerbuddy.io), to help with service discovery and bootstrap orchestration
+- [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform, including the new [Docker log drivers](https://www.joyent.com/blog/docker-log-drivers) support.
+- Nginx, acting as a source of logs for testing.
```
Poor Nginx gets no link. sadface.

misterbisson commented on README.md:
````diff
@@ -1,76 +1,106 @@
-### Running the test rig
+```bash
+curl -O https://raw.githubusercontent.com/joyent/sdc-docker/master/tools/sdc-docker-setup.sh && chmod +x sdc-docker-setup.sh
+./sdc-docker-setup.sh -k us-east-1.api.joyent.com <ACCOUNT> ~/.ssh/<PRIVATE_KEY_FILE>
````
I think you're missing a closing fence right below here.

tgross replied:
:+1: will fix

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,242 @@
+# triton CLI
+getIpPort() {
+    if [ -z "${COMPOSE_FILE}" ]; then
+        local ip=$(triton inst get ${COMPOSE_PROJECT_NAME}_$1_1 | json -a ips.1)
```
No desire to use the CNS name?

tgross replied:
I can do that... just need to add CNS labels to Nginx. Probably will change the function name to `getPublicUrl` or something then (I want it to support local development too).

misterbisson commented on README.md:
```diff
@@ -1,76 +1,106 @@
+- [Triton](https://www.joyent.com/), Joyent's container-native infrastructure platform, including the new [Docker log drivers](https://www.joyent.com/blog/docker-log-drivers) support.
+- Nginx, acting as a source of logs for testing.
 
 ![Diagram of Triton-ELK architecture](./doc/triton-elk.png)
```
There's something about this diagram that makes me think that Logstash is a back-end for Nginx. That interpretation isn't easily justified, as the relationship lines clearly do not always indicate "back-end," as the lines to Consul should make clear, it's just that seeing Nginx without a back-end is so unusual.

tgross replied:
Cluttering it with something that we're not supplying here doesn't seem like an awesome idea either though. What if we just mark those arrows with the label "logs"?

misterbisson replied:
Labeling the arrows would definitely help, but then I got concerned that we might need to label all the arrows. 

I also wondered if the diagram made Nginx appear to be too much a part of the ELK composition, and maybe the two should be segregated in separate groups: box around ELK components, box around log generating components.

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,242 @@
+        echo
+        tput rev  # reverse
+        tput bold # bold
+        echo 'Docker is required, but does not appear to be installed.'
```
Don't forget to `tput sgr0 # clear` after this.

tgross replied:
Good catch.

misterbisson commented:
Perhaps this needs a mention of the `_env` file and why it's there?

_And..._ perhaps we should give the user a little help in gearing the `_env` file?

``` bash
local account_uuid=$(triton account get | awk -F": " '/id:/{print $2}')
local target_dc=$(triton profile get | awk -F"/" '/url:/{print $3}' | awk -F'.' '{print $1}')

CONSUL=consul.svc.${account_uuid}.{target_dc}.cns.joyent.com
```

tgross commented:
> Perhaps this needs a mention of the _env file and why it's there?
> And... perhaps we should give the user a little help in gearing the _env file?

Good call on both.

tgross committed:
Use Triton CNS for test script and force check

Use Triton CNS for getting web interfaces of components. Added CNS labels to
Nginx test rig to support that.

Forces check function to run for all top-level functions in the test.sh script.
Updates the check function to write-out the _env file for CONSUL env var and
make TRITON_* env vars available to the rest of the test script.

Includes some minor README fixes

misterbisson commented:
> gearing

Damn you autocorrect!

tgross committed:
Rework architecture diagram

tgross commented:
Updated that diagram so I think this is ready for one more pass.

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,251 @@
+# default behavior
+echo "Starting example application"
+echo "project prefix:      $COMPOSE_PROJECT_NAME"
+echo "docker-compose file: $COMPOSE_FILE"
```
`$COMPOSE_FILE` is empty for the `./test.sh` with no args case, resulting in empty output.

tgross replied:
That's how we've done most of these... if you don't pass COMPOSE_FILE then Compose picks the `docker-compose.yml` by default. I can have it write out the value here instead if you think it'll be less confusing.

misterbisson replied:
Yeah, I'm feeling we should either emit the file we're using here or conditionally not emit the static text label.

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,251 @@
+
+run() {
+    check
+    docker-compose up -d \
```
Consul is missing from this list.

Does it make sense to name each service? Or is a `docker-compose up -d` enough?

misterbisson replied:
It's likely not necessary (given that the `elasticsearch` node can provide the data service), but on testing it seems `elasticsearch_data` is missing from the list as well.

tgross replied:
Yah, that was left over from the original version where we had the Nginx stuffed into the same `docker-compose.yml` file. Will fix.

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,251 @@
+        tput bold # bold
+        echo 'Docker is required, but does not appear to be installed.'
+        echo 'See https://docs.joyent.com/public-cloud/api-access/docker'
+        tput sgr0 # clear
```
Minor inconsistency: the other error messages just highlight one line, not the second information line

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,251 @@
+    done
+    echo
+    echo "$3"
+    open "$1"
```
Do you want to test for `open`, so we don't emit errors here when not on a Mac?

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,251 @@
+getPublicUrl() {
+    if [ -z "${COMPOSE_FILE}" ]; then
+        local cname=$(printf '%s.svc.%s.%s.triton.zone' $1 ${TRITON_ACCOUNT} ${TRITON_DC})
+        echo "$cname:$port"
```
`$port` is `$2`?

tgross replied:
This should be: `echo "$cname:$2"`

tgross committed:
Usability improvements and bug fixes for test.sh

- Check that `open` works before opening web consoles
- Bring up all services in `run`, not by name
- Fix missing port in getPublicUrl when we are using CNS

tgross commented:
I've picked up those comments in https://github.com/tgross/triton-elk/pull/2/commits/bf24c3526bbdcdc3866887a41c0154ce2042be84

tgross committed:
Use triton.zone for Consul CNS domain

tgross committed:
Split default index setup out into the Kibana health check

tgross committed:
Make sure netcat exits when we initialize log entries

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,255 @@
+        exit 1
+    fi
+
+    echo CONSUL=consul.svc.${TRITON_ACCOUNT}.${TRITON_DC}.triton.zone > _env
```
I'm agreeing that it is very strange this did not work as `...cns.joyent.com`, as the private name should be usable by everything on the same fabric.

misterbisson replied:
We learned that one of the DCs did not match the expected pattern. That's changed or changing. Should we switch this back?

tgross replied:
Let's avoid changing it until after it's done, so that we don't get folks from inside Joyent filing bugs in the meantime.

tgross committed:
Add CNS label to Kibana and Logstash

tgross committed:
Don't run tests by default because we can't capture type arg

misterbisson commented on nginx/containerbuddy.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "consul": "consul:8500",
+  "consul": "{{ .CONSUL }}:8500",
   "onStart": "/opt/containerbuddy/reload.sh",
```
This calls https://github.com/tgross/triton-nginx/blob/master/bin/reload.sh#L19, causing it to try `nginx -s reload` before the Nginx starts, triggering an error:

```
2016/03/17 18:53:53 [error] 11366#11366: open() "/var/run/nginx.pid" failed (2: No such file or directory)
nginx: [error] open() "/var/run/nginx.pid" failed (2: No such file or directory)
```

tgross replied:
Noted in https://github.com/tgross/triton-nginx/issues/6 and https://github.com/tgross/workshop-autopilot/issues/2. We'll need to update in the Docker image and here in the Containerbuddy config.

misterbisson commented on test.sh:
```diff
@@ -0,0 +1,254 @@
+    esac
+
+    echo 'Starting Nginx log source...' && \
+        CONSUL=$(getPrivateIpPort consul 8500 tcp) \
```
Does it make sense to get `CONSUL` from here, or should we use the value from `_env`? The inconsistency seems unnecessary.

tgross replied:
When running locally we need to use links and not CNS, so the `_env` file which gets used by `docker-compose.yml` is populated by CNS while the `local-compose.yml` file just uses links and the static string "consul". If all we wanted to do was stand up some containers we could just do all this logic in two compose files, but we're doing things like opening the web interfaces.

~~You've asked about this a couple different ways now, so let me propose this approach:~~
~~\- we ditch the `test.sh` file entirely~~
~~\- we provide a `docker-compose.yml`, `local-compose.yml`, `test-compose.yml`, and `test-local-compose.yml` that does all the standing up of things.~~
~~\- we provide a `check.sh` file to do the environment checking and open web portals, but it only works on Triton and not locally.~~

Nevermind, we have to inject the Nginx config and Containerbuddy config into the `0x74696d/triton-nginx` container, which we can't do via Compose.

misterbisson replied:
This is a-ok as is.

tgross committed:
Remove redundant port number from Nginx CB config

tgross committed:
Remove test-compose to workaround env var interpolation bug in Compose

Compose 1.6.? has a bug where our LOGSTASH variable is not being interpolated
and so we end up just using the json driver because it's unpopulated. This
works around the problem by just calling out to `docker run`

tgross committed:
Use preStart arg on nginx reload.sh

misterbisson commented:
:house_with_garden: :walking:
